### ISSUE TITLE
Positivity checker marks postulates as constant in mutual block

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3404]
The implicit argument of `f` at the last line of the code below is not solved by Agda (it should be). Note that if you remove `x'` (which isn’t used anywhere), then it does get solved.

```agda
postulate
  X : Set
  x : X
  A : Set

x' : X
postulate
  B : A → Set
x' = x

postulate
  f : {x : A} → B x

g : {x : A} → B x
g = f
```

**Comment by guillaumebrunerie** [comment]
The following is type-checked with no error. This doesn’t look good.
```agda
g : {a a' : A} → B a → B a'
g x = x
```

**Comment by guillaumebrunerie** [comment]
Does that qualify for the `false` label? There are postulates, but it’s simply the existence of the 2^ function, and its values at 0 and 1.

```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

module _ where

_∙_ : {A : Set} {a b c : A} → a ≡ b → b ≡ c → a ≡ c
refl ∙ refl = refl

! : {A : Set} {a b : A} → a ≡ b → b ≡ a
! refl = refl

x' : Set
postulate
  2^ : Nat → Nat
  2^0 : 2^ 0 ≡ 1
  2^1 : 2^ 1 ≡ 2
x' = Nat

bad : 1 ≡ 2
bad = (! 2^0) ∙ 2^1
```

**Comment by UlfNorell** [comment]
I think you deserve the `false` label for this :-)

Got rid of implicit arguments (to rule out elaboration problems) and some minor simplifications:
```agda
postulate
  X : Set
  A : Set

Y : Set
postulate B : A → Set
Y = X

postulate
  f : (a : A) → B a

g : (a : A) → B a
g a = f _  -- yellow
```

**Comment by andreasabel** [comment]
More concisely, use old-style `mutual` blocks:
```agda
postulate
  A : Set

mutual
  postulate B : A → Set

postulate
  f : (a : A) → B a

g : (a : A) → B a
g a = f _  -- yellow
```
I do not understand what it going on here.  Why should there be unsolved metas from the `mutual` block?

Old-style `mutual` makes also the `false` example shorter:
```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

trans : {A : Set} {a b c : A} → a ≡ b → b ≡ c → a ≡ c
trans refl refl = refl

mutual
  postulate
    2^  : Nat → Nat
    2^0 : 1 ≡ 2^ 0
    2^1 : 2^ 1 ≡ 2

bad : 1 ≡ 2
bad = trans 2^0 2^1
```

**Comment by andreasabel** [comment]
Apparently Agda thinks that `B` is a constant function (and, likewise `2^`).
The positivity checker seems to treat postulates in mutual blocks as functions with an empty body, thus, of course, none of its arguments are used, and it is declared a constant function.

Agda-2.5.3 still complains about `1 != 0 of type Nat`, but Agda-2.5.4.1 accepts the false program.

**Comment by andreasabel** [comment]
Assuming postulates to be constant is maybe not in the intention of the user, but does not introduce inconsistency.
Here is your actual inconsistency, using primitives:
```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

cong : ∀{A B : Set} (f : A → B) {x y : A} (eq : x ≡ y) → f x ≡ f y
cong f refl = refl

{-# BUILTIN WORD64 Word64 #-}

primitive        primWord64ToNat   : Word64 → Nat
mutual primitive primWord64FromNat : Nat → Word64
  -- Before fix of #3404, positivity checker judged
  -- a primitive in a mutual block as constant.

fromNatConst : ∀ x y → primWord64FromNat x ≡ primWord64FromNat y
fromNatConst x y = refl

0≡1 : 0 ≡ 1
0≡1 = cong primWord64ToNat (fromNatConst 0 1)
```

**Comment by asr** [comment]
The fix hasn't been back-ported to 2.5.4.3.

**Comment by asr** [comment]
Back-ported into 2.5.4.3.
